TW 372: add CLI installation to CLI ref index (#1970)
temporalio/documentation: docs-src/concepts/what-is-the-temporal-cli.md

This documentation reflects the release of the [single-binary Temporal CLI](https://github.com/temporalio/cli).
This tool was formerly known as `tctl next`.

---

> The Temporal CLI is a command-line tool that serves as a distribution of the [Temporal Server](/concepts/what-is-the-temporal-server) and [Temporal Web UI](/concepts/what-is-the-temporal-web-ui).

Reviewer: ```suggestion
The Temporal CLI is a command-line tool that includes as a distribution of a Temporal Cluster ([Temporal Server](/concepts/what-is-the-temporal-server) , persistence (SQLite) and the [Temporal Web UI](/concepts/what-is-the-temporal-web-ui)).
```

- [temporal server](/cli/server)
- [temporal task-queue](/cli/task-queue/)
- [temporal workflow](/cli/workflow/)

## Installation

> Temporal CLI can be installed through several different methods. While most of them can be used across all operating systems, please note that Homebrew is macOS-exclusive.

Reviewer 2: @DJSanti the note about macos is irrelevant in this context.

For more information, see our guide to [running a Development Cluster](/clusters/how-to-install-temporal-cli).

## Starting the Temporal Server

Run the following command to start the Temporal Server.
This command also starts the Web UI.

```bash
temporal server start-dev
```

At this point you should have a server running on `localhost:7233` and a web interface at <http://localhost:8233>.
A default Namespace has also been created.

Run individual commands to interact with the local Temporal Server.

```bash
temporal operator namespace list
temporal workflow list
```

## Configuration

> Use the help flag to see a full list of CLI options:

Reviewer 2: From the installation guide:

> The server's startup configuration can be customized using command line options. For a full list of options, run:

Reviewer 2: I'd also move this section closer to the `Starting the Temporal Server` header.

| `TEMPORAL_CLI_TLS_SERVER_NAME`               | Provides an override for the target TLS server name.                              | --tls-server-name               |

## Auto-completion

The Temporal CLI has the capability to auto-complete commands.

Running `temporal completion SHELL` outputs the related completion SHELL code.

### zsh auto-completion

<!-- TODO: add more information about zsh to make comparable to bash section -->

Add the following code snippet to your `~/.zshrc` file:

```sh
source <(temporal completion zsh)
```

If you're running auto-completion from the terminal, run the following command:

```sh
echo 'source <(temporal completion zsh)' >> ~/.zshrc
```

After setting the variable, run the following command:

`source ~/.zshrc`.

### Bash auto-completion

Bash auto-completion relies on `bash-completion`.

Install the software with the steps provided in the [bash-completion README](https://github.com/scop/bash-completion#installation), or use your preferred package manager on your operating system.

> To view different bash versions, go [here](https://repology.org/project/bash-completion/versions).

Reviewer 2: Should show how to install the completion like we do with zsh completion.

> To view different bash versions, go [here](https://repology.org/project/bash-completion/versions).

These aren't different bash versions.

- [temporal server](/cli/server)
- [temporal task-queue](/cli/task-queue/)
- [temporal workflow](/cli/workflow/)

> ## Installation

Reviewer 2: Please remove the installation instructions if you're linking to the reviewed guide.

Temporal CLI can be installed through several different methods. While most of them can be used across all operating systems, please note that Homebrew is macOS-exclusive.

For more information, see our guide to [running a Development Cluster](/clusters/how-to-install-temporal-cli).

### cURL

Run the following command to install the latest version of Temporal CLI.

`curl -sSf https://temporal.download/cli.sh | sh`

### Homebrew

Run the following command in a new terminal window to install CLI for macOS.

`brew install temporal`

### GitHub releases

Download and extract the latest release from [GitHub releases](https://github.com/temporalio/cli/releases).

### CDN

To install the Temporal CLI from CDN:

1. Select the binary with the corresponding platform and architecture of your system.
2. Download the binary.
3. Extract the downloaded archive.
4. Add the `temporal` binary to your system PATH.

> ## Starting the Temporal Server

Reviewer: ```suggestion
## Starting a local dev Cluster
```

You will want to update this in source obviously :)

Reviewer: And I know that using Cluster will be controversial to @bergundy - but embracing consistency with our terms is pretty important - and this consistency needs to apply to the CLI repo.
Since the current owners of that repo are failing to commit even after disagreeing... then we need to update this for them.

Reviewer 2: You'll want to update the server repo too:
https://github.com/temporalio/temporal#download-and-start-temporal-server-locally
And docker compose:
https://github.com/temporalio/docker-compose#temporal-server-docker-compose-files
And pretty much every other repository that mentions Temporal server.

Run the command provided below to start the Temporal Server.
This will automatically start the Web UI.

```bash
temporal server start-dev
```

At this point you should have a server running on `localhost:7233` and a web interface at <http://localhost:8233>.
A default Namespace has also been created.

Run individual commands to interact with the local Temporal server.

```bash
temporal operator namespace list
temporal workflow list
```

## Configuration

Use the help flag to see a full list of CLI options:

```bash
temporal server start-dev --help
```

> ### Environmental variables

Reviewer: Environment variables apply to the whole CLI - not just the `server start-dev` command.

Reorg the content appropriately - this is sitting right between two `server` command info bits.

| `TEMPORAL_CONTEXT_TIMEOUT`                   | An optional timeout for the context of an RPC call (in seconds). Default value: 5 | --context-timeout               |
| `TEMPORAL_CLI_NAMESPACE`                     | Identifies a Namespace in the Temporal Workflow. Default value: "default".        | --namespace                     |
| `TEMPORAL_CLI_TLS_CA`                        | Path to server CA certificate.                                                    | --tls-ca-path                   |
| `TEMPORAL_CLI_TLS_CERT`                      | Path to x509 certificate.                                                         | --tls-cert-path                 |
| `TEMPORAL_CLI_TLS_DISABLE_HOST_VERIFICATION` | Disables TLS host name verification if already enabled. Default value: false.     | --tls-disable-host-verification |
| `TEMPORAL_CLI_TLS_KEY`                       | Path to private certificate key.                                                  | --tls-key-path                  |
| `TEMPORAL_CLI_TLS_SERVER_NAME`               | Provides an override for the target TLS server name.                              | --tls-server-name               |

> ### Namespace registration

Reviewer: Can we not now link to the CLI reference for this?

I don't think the README should break down every high level command usage.

This is the stuff that needs to go into the reference itself.

Having the list of "available commands" near the top will suffice.

Reviewer: Actually - this makes sense in the context of running the dev Cluster.

Namespaces are pre-registered at startup so they're available to use right away.
To customize the pre-registered namespaces, start the server with:

```bash
temporal server start-dev --namespace foo --namespace bar
```

You can also register Namespaces with the following command:

```bash
temporal operator namespace create foo
```

### Persistence modes

By default, `temporal server start-dev` runs in an in-memory mode.

To persist the state to a file on disk, use `--db-filename`:

```bash
temporal server start-dev --db-filename my_test.db
```

### Enable or disable Temporal UI

By default, the Temporal UI is started with Temporal CLI. The UI can be disabled via a runtime flag:

```bash
temporal server start-dev --headless
```

### Dynamic configuration

Advanced configuration of the Temporal CLI requires the use of a dynamic configuration file.
This file is created outside of the Temporal CLI; it is usually located with the service's config files.

Dynamic configuration values can also be set via `--dynamic-config-value KEY=JSON_VALUE`.

> For example, to enable the search attribute cache, run:

Reviewer 2: ```suggestion
For example, to enable the search attribute cache (disabled by default), run:
```

### zsh auto-completion

<!-- TODO: add more information about zsh to make comparable to bash section -->

Add the following code snippet to your `~/.zshrc` file:

```sh
source <(temporal completion zsh)
```

If you're running auto-completion from the terminal, run the command below:

```sh
echo 'source <(temporal completion zsh)' >> ~/.zshrc
```

After setting the variable, run:

`source ~/.zshrc`.

### Bash auto-completion

Bash auto-completion relies on `bash-completion`.

Install the software with the steps provided [here](https://github.com/scop/bash-completion#installation), or use your preferred package manager on your operating system.

#### Linux installation

Verify that `bash-completion` is installed by running `type _init_completion`.

To install the software on Alpine Linux, run:

```bash
apk update
apk add bash-completion
source /etc/profile.d/bash_completion.sh
```

Finally, [enable completion for Temporal in your script](#auto-completion).

#### macOS installation

Install `bash-completion` through Homebrew:
`brew install bash-completion@2`

Add the provided code snippet to your `~/.bashrc` file:

> 

Reviewer 2: Link to the bash completion installation guide here instead of providing our own instructions.

Reviewer 2: Can add this in addition to the GH link: https://repology.org/project/bash-completion/versions
